keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+            * [Non-generic TTL-based Garbage Collection](#non-generic-ttl-based-garbage-collection)
+            * ["Root" Resources Without an Owner](#root-resources-without-an-owner)
+            * [Immutable Pattern: Short-lived, Frequently-created Resources](#immutable-pattern-short-lived-frequently-created-resources)
+            * [Cannot Centralize Cleanup of Multiple Resources with OwnerReferences](#cannot-centralize-cleanup-of-multiple-resources-with-ownerreferences)
+      * [Proposal](#proposal)
+         * [Generic TTL-based Garbage Collection](#generic-ttl-based-garbage-collection)
+         * [Clean Up Resources After Use](#clean-up-resources-after-use)
+         * [Summary](#summary-2)
+         * [User Stories](#user-stories)
+            * [Story 1: Clean up old Jobs](#story-1-clean-up-old-jobs)
+            * [Story 2: Secrets and ConfigMaps](#story-2-secrets-and-configmaps)
+            * [Story 3: CustomResourceDefinitions (CRDs) and custom resources](#story-3-customresourcedefinitions-crds-and-custom-resources)
+         * [Implementation Details](#implementation-details)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+      * [Future Work](#future-work)
+         * [Generic Capacity-based Garbage Collection](#generic-capacity-based-garbage-collection)
+         * [Namespace-wide Garbage Collection Policy](#namespace-wide-garbage-collection-policy)
+         * [Move Event TTL out of etcd TTL](#move-event-ttl-out-of-etcd-ttl)
+         * [Make Pod GC and Event GC Work Better with Garbage Collector](#make-pod-gc-and-event-gc-work-better-with-garbage-collector)
+         * [Move deletionGracePeriodSeconds from metadata to Pod spec](#move-deletiongraceperiodseconds-from-metadata-to-pod-spec)
+      * [Alternatives](#alternatives)
+
+## Summary
+
+We propose to expand Kubernetes garbage collection to handle common uses which
+are awkward today. We will add an object-lifecycle-respecting TTL mechanism to
+the garbage collector. We will also allow users to mark resource objects to be
+cleaned up by the garbage collector after those objects are no longer in use:
+either terminated, completed, or not referenced by other resource objects.
+
+## Motivation
+
+Make it easy for users to clean up frequently-created, short-lived Kubernetes
+resources objects.
+
+One common use case is [ConfigMaps][] and [Secrets][]. The users roll out
+ConfigMaps and Secrets by creating new ConfigMaps or Secrets and updating the
+references to those ConfigMaps or Secrets in containers ([#22368][]). The old
+ConfigMaps and Secrets, if not referenced by any container, are no longer
+needed. 
+
+Another common use case is completed [Jobs][]. Users keep creating Jobs to run
+some batch tasks, and those Jobs, once complete, are no longer needed in the
+system.
```

Comment:
@soltysh might prove me wrong, but I thought there is something like revisionHistoryLimit for Jobs which will cause the old jobs to be deleted by the jobs controller.

Reply:
Here I'm talking about the Jobs that aren't managed by CronJobs. Some users create Jobs directly with their own tools.